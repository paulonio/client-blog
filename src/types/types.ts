export interface Post {
  username: string;
  date: string;
  title: string;
  body: string;
  urlToImage: string;
}
