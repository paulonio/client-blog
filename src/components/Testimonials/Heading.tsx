import React from 'react';
import { useTranslation } from 'next-i18next';

import { sen } from '@/styles/fonts';

import styles from './styled.module.scss';

const Heading = () => {
  const { t } = useTranslation('testimonials');

  return (
    <div className={styles.heading}>
      <h5>{t('heading')}</h5>
      <h2 className={sen.className}>{t('title')}</h2>
      <p className="body-s">{t('body')}</p>
    </div>
  );
};

export default Heading;
