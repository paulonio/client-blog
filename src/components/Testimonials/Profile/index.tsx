import React, { FC } from 'react';
import { useTranslation } from 'next-i18next';

import type { User } from '@/types/types';

import styles from './styled.module.scss';
import { stringToKey } from '@/utils/utils';

interface ProfileProps {
  profile: User;
}

const Profile: FC<ProfileProps> = ({ profile }) => {
  const { fullName, urlToAvatar, city, country } = profile;
  const { t } = useTranslation('common');

  return (
    <div className={styles.profile}>
      <div className={styles.avatar}>
        <img src={urlToAvatar} alt={fullName} />
      </div>
      <div>
        <h4>{t(stringToKey(fullName))}</h4>
        <p className="body-s">
          {t(stringToKey(city))}, {t(stringToKey(country))}
        </p>
      </div>
    </div>
  );
};

export default Profile;
