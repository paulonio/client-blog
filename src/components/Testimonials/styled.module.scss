@import '../../styles/sass/constants';

.heading {
  width: 33%;
  display: flex;
  flex-direction: column;
  align-items: start;
  row-gap: 12px;

  & h2 {
    margin-bottom: 4px;
  }
}

.testimonial {
  margin-top: 128px;
  padding: 80px 112px 75px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 80px;
  background-color: $color-light-s;

  &__wrapper {
    width: 59%;
    padding-left: 96px;
    display: flex;
    flex-direction: column;
    row-gap: 121px;
    border-left: solid 1px rgba($color-dark-s, 0.4);

    & h4 {
      padding-right: 62px;
    }
  }

  &__footer {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }
}

.profile {
  display: flex;
  align-items: center;
  justify-content: start;
  gap: 16px;

  &__wrapper {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }
}

.avatar {
  width: 48px;
  height: 48px;
  border-radius: 50px;
  overflow: hidden;
}

.buttons {
  display: flex;
  align-items: center;
  justify-content: end;
  gap: 24px;

  & button {
    width: 62px;
    height: 62px;
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: 50px;
    background-color: $color-black;
    border: none;

    & svg {
      fill: $color-light-m;
    }
  
    &:disabled {
      width: 48px;
      height: 48px;
      background-color: $color-white;
      & svg {
        fill: $color-black;
      }
    }
  }
}