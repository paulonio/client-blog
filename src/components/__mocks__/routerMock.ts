import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

(useRouter as jest.Mock).mockReturnValue({ query: {} });
