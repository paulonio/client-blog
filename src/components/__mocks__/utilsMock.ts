import { isLinkActive, parseString } from '@/utils/utils';

jest.mock('../../utils/utils', () => ({
  isLinkActive: jest.fn(),
  parseString: jest.fn(),
}));

(isLinkActive as jest.Mock).mockReturnValue(true);
(parseString as jest.Mock).mockReturnValue('');
