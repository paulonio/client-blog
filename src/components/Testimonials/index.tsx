'use client';

import React, { FC, useRef, useState } from 'react';
import { useTranslation } from 'next-i18next';

import Heading from './Heading';
import Profile from './Profile';
import ArrowIcon from '@/icons/ArrowIcon';
import Loader from '@/components/Loader';

import { Testimonial } from '@/types/types';
import { useIntersectionObserver } from '@/hooks/useIntersectionObserver';
import { stringToKey } from '@/utils/utils';

import styles from './styled.module.scss';

interface TestimonialsProps {
  testimonials: Testimonial[];
}

const Testimonials: FC<TestimonialsProps> = ({ testimonials }) => {
  const [index, setIndex] = useState<number>(0);
  const lastIndex = testimonials.length - 1;
  const ref = useRef<HTMLElement | null>(null);
  const { isShowing, isLoading } = useIntersectionObserver(ref);
  const { t } = useTranslation('testimonials');

  const handleIncrement = () => {
    if (index !== lastIndex) {
      setIndex((idx) => idx + 1);
    }
  };

  const handleDecrement = () => {
    if (index !== 0) {
      setIndex((idx) => idx - 1);
    }
  };

  return (
    <section ref={ref} className={styles.testimonial}>
      {isLoading && <Loader />}
      {isShowing && (
        <>
          <Heading />
          <div className={styles.testimonial__wrapper}>
            <div style={{ marginLeft: `${-index * 100}%` }} className={styles.slides}>
              {testimonials.map(({ profile, message }) => {
                return (
                  <div key={profile.fullName} className={styles.slide}>
                    <h4>{t(stringToKey(message))}</h4>
                    <div className={styles.testimonial__footer}>
                      <Profile profile={profile} />
                      <div className={styles.buttons}>
                        <button
                          type="button"
                          disabled={index === 0}
                          className={styles.prev}
                          onClick={handleDecrement}
                        >
                          <ArrowIcon className={styles.prev__icon} />
                        </button>
                        <button
                          type="button"
                          disabled={index === lastIndex}
                          className={styles.next}
                          onClick={handleIncrement}
                        >
                          <ArrowIcon className={styles.next__icon} />
                        </button>
                      </div>
                    </div>
                  </div>
                );
              })}
            </div>
          </div>
        </>
      )}
    </section>
  );
};

export default Testimonials;
