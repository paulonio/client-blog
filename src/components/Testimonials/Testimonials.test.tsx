import React from 'react';
import { render } from '@testing-library/react';
import Testimonials from '.';
import { Testimonial } from '@/types/types';
import '../__mocks__/intersectionObserverMock';

const TESTIMONIALS_TEST: Testimonial[] = [
  {
    profile: { urlToAvatar: 'test', fullName: 'test', city: 'test', country: 'test' },
    message: 'test',
  },
];

describe('Special', () => {
  it('should render special', () => {
    const { container } = render(<Testimonials testimonials={TESTIMONIALS_TEST} />);

    expect(container).toMatchSnapshot();
  });
});
