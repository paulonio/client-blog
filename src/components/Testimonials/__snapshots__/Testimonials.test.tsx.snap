// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Special should render special 1`] = `
<div>
  <section
    class="testimonial"
  />
</div>
`;
